Prevent assigning class name to variable.

errael opened the pull request:
@lifepillar @yegappan @dkearns please review.
Code reviewers, please consider where the errors are generated. Is there a better place to check? Are all the cases caught?


Error is `E1393: Cannot use a class as a variable or value`

**notes on changes**

Prevent assignment
- from non-value items, e.g. `class` or `typealias`.
- of non-value items as function argument or return.
- to `dict`/`list`: `var l = [1, C]`, `var d = {x: C}`.

NOTE: currently non-value items can be added to a list (while #13421 is being sorted out)
NOTE: a builtin must not return a non-value:, bare, in `list`, in `dict`

Here's an overview of the code changes made (so far).
```
simple assignments
    scriptlevel assignment  ex_let_one in evalvars.c
    :def assignment         compile_assignment in vim9compile.c

Function related checks
    scriptlevel argument    get_func_arguments in userfunc.c
    argument, call in def   check_ufunc_arg_types in vim9execute.c
    return value            ISN_RETURN (but gives wrong line number)
    defer/funcref           check_func_args_from_type in vim9instr.c

dict/list checks
    script list         eval_list in list.c
    script dict         eval_dict in dict.c
    compile list/dict   get_member_type_from_stack in vim9instr.c
```
There is a separate commit for `instanceof()` experimentation.
There is a separate commit for uniform builtin argument checking.

**TODO**
- More tests.
  Fixed tests as needed, but have around a dozen or so tests to write based on debug scripts. Plan to add tests after initial review (particularly of the error message).

**What now?**

This PR is very conservative. Some of the checks could be backed out.
The advantage of aggressive checks is uniformity, which should be less
confusing.
- No extra builtin checks, builtins have their own checks.
- Allow non-values in list/dict.
  This would ease some problems with builtins like `instanceof()`
  This still wouldn't allow the use of the non-value.
  Need to prevent non-value from being the result of an expression;
  protects against usage like aList[0].new().


---

**Notes on why the change**

With this PR, `Test_call_constructor_from_legacy()` failed, the test has been
replaced with `Test_construct_object_from_legacy()` which shows an alternate
way to do the same thing.

There's the following TODO item:
```
When "Meta" is a class, is "const MetaAlias = Meta" allowed?  It should
either work or given an error. Possibly give an error now and implement it
later (using a typedef).  #12006
```

This PR "give[s] an error now".
The important question: does preventing this take away an important capability?

When I first saw this TODO, I wondered why not just let it be, and take care
of it later. I played with it, #12961, and noted that it was very limited with
common stuff not working. I think I didn't read the original issue, #12006,
carefully; just reading it now, Bram's argument is
- it should be handled by typedef whenever something like that becomes available
- there is no class type.

If it's left in, **it leaves a lot of technical debt** to get it fully working.

Some uses for this, can be achieved with something like
```
class C
    static def CreateC(): any ... enddef
endclass
var CreateC: func = C.Create
```

errael committed:
Prevent assigning class name to variable.

yegappan commented on src/evalvars.c:
```diff
@@ -1158,7 +1158,9 @@ ex_let(exarg_T *eap)
     SOURCING_LNUM = cur_lnum;
 
-    if (!eap->skip && eval_res != FAIL)
+    if (!eap->skip && eval_res != FAIL && rettv.v_type == VAR_CLASS)
```
You can use list assignment to assign multiple variables (B is a class in the statement below):

```
    [a, b] = [10, B]
```

To handle this case, this check should be moved to the ex_let_one() function.

errael replied:
> To handle this case, this check should be moved to the ex_let_one() function

Thanks, it was still on my list.

errael committed:
Modify `instanceof()` to handle varargs of classes instead of list<any>.

errael committed:
Uniform builtin function argument checking.
This may not be needed. The builtins check their args.
This commit adds uniformity to checking.

errael committed:
Additional erge cleanup.
